proposals/0448-type-variable-scoping.rst

```diff
@@ -0,0 +1,1167 @@
+   and is additionally available in pattern synonym pattern bindings (where it
+   brings into scope only universals) and pattern synonym expression bindings
+   (where is brings into scope both universals and existentials). (In an implicitly
+   bidirectional pattern synonym, the ``@(..)`` brings into scope only universals.)
+
+   **Discussion**
+
+   This new notation seems like a convenient middle ground,
+   allowing for an easy transition from the old-style ``-XScopedTypeVariables``
+   to the newer ``-XTypeAbstractions``. It brings the *inferred* variables (from `#99`_)
+   into
+   scope, quite conveniently. This new notation also allows type variables to
+   be brought into scope without the ``forall`` keyword in the type, in case
+   the user does not want to trigger ``forall``\ -or-nothing behavior.
+
+   Note that this notation is forward compatible with visible dependent quantification
+   in terms (`#281`_)::
+
+     f :: foreach (count :: Int) (label :: String) (is_paid_for :: Bool) -> Invoice
+     f (..) = -- here, count, label, and is_pair_for are all in scope
+
+   This style allows for more perspicuous types while avoiding redundancy. The particular
+   example here uses ``foreach`` to denote arguments that are available at runtime, but
+   nothing about ``foreach`` is required to make this all work (as far as scoping is
+   concerned).
+
+   Accepting the ``@(..)`` syntax does *not* entail accepting this new, separate
+   ``(..)`` syntax, though it is good to know that the idea is forward compatible.
+
+   A ``@(..)`` argument counts as a type argument when asking whether ``-XScopedForAlls``
+   affects a function equation.
+
+   The new ``@(..)`` notation does *not* work with expression type signatures,
+   lambda-expressions, or anywhere other than a function binding with a type
+   signature. This is because doing so would require propagating type
+   information into scoping, which is problematic.
+
+   Some have argued on GitHub that it may be best to hold off the ``@(..)`` until
+   we gain more experience here: adding new features is easier than removing them.
+   While I agree that this could be done, the ``@(..)`` construct makes for a very
+   easy migration from today's ``-XScopedTypeVariables`` and is thus tempting to
+   be around from the start. I don't feel strongly.
+
+#. We could simply make ``-XScopedForAlls`` and ``-XTypeAbstractions`` incompatible.
+   If the user specifies both, reject the program.
+
+   I find this approach less convenient, as it prevents an easy migration from the
+   status quo (with ``-XScopedTypeVariables`` enabled often, including in ``-XGHC2021``)
+   to a future relying more on ``-XTypeAbstractions``. The approach described in this
+   proposal means that enabling ``-XTypeAbstractions`` affects nothing about ``-XScopedForAlls``,
+   until a user tries to actually use a type abstraction. That's a nice property.
+
+``let``-binding types
+~~~~~~~~~~~~~~~~~~~~~
```

Comment:
I agree with @simonpj . The subproposals about let are require more discussion than the rest. They'll be best served as a separate proposal (which I'll be happy to shepherd as well :slightly_smiling_face: ). Keeping it in this proposal will only result in slowing down the conversation.